hamp_radar/serde_flight.py

```diff
@@ -0,0 +1,160 @@
+
+    Raises:
+        ValueError: If there are two consecutive PPAR subblocks, indicating corrupt data.
+    """
+    if len(mmbg.subblocks) > 1:
+        raise ValueError(
+            "Two consecutive PPAR subblocks should never occuur, corrupt data?"
+        )
+
+    if mmbg.count > 1:
+        # only use last PPAR for DSP configuration
+        offset = mmbg.offset + mmbg.step * (mmbg.count - 1)
+        count = 1
+        mmbg = MultiMainBlockGeometry(
+            mmbg.tag, offset, count, mmbg.step, mmbg.subblocks
+        )
+
+    ppar = DatasetBlockGeometry(filename, mmbg)
+    return ppar
+
+
+def convert_to_datasetgeometries(
+    pfgs: Iterable[PdsFileGeometry],
+) -> List[DatasetGeometry]:
+    """
+    Converts an iterable of PdsFileGeometry instances into a list of DatasetGeometry instances.
+
+    This function converys a collection of PdsFileGeometry instances into
+    DatasetGeometry instances. It identifies "PPAR" tags to start new datasets and appends
+    data blocks to the most recent dataset. If consecutibe PPAR tags are found withou
+    data in between, the ppar of the current dataset is updated to match the last given PPAR.
+
+    Args:
+        pfgs (Iterable[PdsFileGeometry]): An iterable of PdsFileGeometry instances, e.g. for a flight.
+
+    Returns:
+        List[DatasetGeometry]: A list of DatasetGeometry instances, e.g. for each dataset in a flight.
+
+    Raises:
+        UserWarning: If data is found before the first "PPAR" tag in the PdsFileGeometry's mainblocks.
+    """
+    import warnings
+
+    datasets, is_ppar = [], False
+    for pfg in pfgs:
+        for mmbg in pfg.mainblocks:
+            if mmbg.subblocks[0].tag == "PPAR":
+                ppar = ppar_for_dataset(pfg.filename, mmbg)
+                if is_ppar:
+                    # update ppar of existing dataset
+                    datasets[-1].ppar = ppar
+                else:
+                    # start new dataset
+                    datasets.append(DatasetGeometry(ppar, []))
+                    is_ppar = True
+            else:
+                if datasets == []:
+                    # skip data if no PPAR has yet been found in pfg.mainblocks
+                    warnings.warn(
+                        "Warning: data found before first PPAR is being skipped",
+                        UserWarning,
+                    )
+                else:
+                    # append to latest dataset
+                    is_ppar = False
+                    datablock = DatasetBlockGeometry(pfg.filename, mmbg)
+                    datasets[-1].data.append(datablock)
+    return datasets
+
+
+def serialize_flight(flightname: str, pfgs: Iterable[PdsFileGeometry]):
+    from serde.json import to_json
+
+    datasets = convert_to_datasetgeometries(pfgs)
+    flight = FlightGeometry(flightname, datasets)
+    return to_json(flight)
+
+
+def write_flight_geometry(
+    flightname: str, pfgs: Iterable[PdsFileGeometry], geomsdir: Path
+):
```

Comment:
I'm wondering if this function probably should just take an output filename instead of a `geomsdir`. It seems that knowning about the existence of the `geomsdir` is a bit too much. Also I'm wondering why the `write_flight_geometry` function doesn't accept a `FlightGeometry`...

Reply:
this function works in the same way as write_pdsfile_geometries. The idea is it's saying "give me a flight and I'll write you its geometry" I therefore thought the name of the file written should be the name of the flight (hence specify where to put the file not the filename itself) and I thought it should take flight pds files / geometries not the fligth geometry which it will figure out for you